extension/js/common/helpers.ts

```diff
@@ -3,8 +3,119 @@
+    }
+    for (const prv of prvs) {
+      await ContactStore.update(undefined, email, { pubkey: KeyUtil.armor(await KeyUtil.asPublicKey(prv)) });
+    }
+  }
+};
+
+const parseAndCheckPrivateKeys = async (decryptedPrivateKeys: string[]) => {
+  const unencryptedPrvs: Key[] = [];
+  // parse and check that all the keys are valid
+  for (const entry of decryptedPrivateKeys) {
+    const { keys, errs } = await KeyUtil.readMany(Buf.fromUtfStr(entry));
+    if (errs.length) {
+      throw new Error(`Some keys could not be parsed`);
+    }
+    if (!keys.length) {
+      throw new Error(`Could not parse any valid keys`);
+    }
+    for (const prv of keys) {
+      if (!prv.isPrivate) {
+        throw new Error(`Key ${prv.id} is not a private key`);
+      }
+      if (!prv.fullyDecrypted) {
+        throw new Error(`Key ${prv.id} is not fully decrypted`);
+      }
+    }
+    unencryptedPrvs.push(...keys);
+  }
+  return { unencryptedPrvs };
+};
+
+const filterKeysToSave = async (candidateKeys: Key[], existingKeys: KeyInfoWithIdentity[]) => {
+  if (!existingKeys.length) {
+    return candidateKeys;
+  }
+  const result: Key[] = [];
+  for (const candidate of candidateKeys) {
+    const longid = KeyUtil.getPrimaryLongid(candidate);
+    const keyToUpdate = existingKeys.filter(ki => ki.longid === longid && ki.family === candidate.family);
+    if (keyToUpdate.length === 1) {
+      const oldKey = await KeyUtil.parse(keyToUpdate[0].private);
+      if (!oldKey.lastModified || !candidate.lastModified || oldKey.lastModified >= candidate.lastModified) {
+        continue;
+      }
+    } else if (keyToUpdate.length > 1) {
+      throw new Error(`Unexpected error: key search by longid=${longid} yielded ${keyToUpdate.length} results`);
+    }
+    result.push(candidate);
+  }
+  return result;
+};
+
+export const processAndStoreKeysFromEkmLocally = async (
+  { acctEmail, decryptedPrivateKeys, options }: { acctEmail: string, decryptedPrivateKeys: string[], options?: PassphraseOptions }
+): Promise<Bm.Res.ProcessKeysFromEkm> => {
+  const { unencryptedPrvs } = await parseAndCheckPrivateKeys(decryptedPrivateKeys);
+  const existingKeys = await KeyStore.get(acctEmail);
+  let passphrase = options?.passphrase;
+  if (passphrase === undefined && !existingKeys.length) {
+    return { unencryptedKeysToSave: [], updateCount: 0 }; // return success as we can't possibly validate a passphrase
+    // this can only happen on misconfiguration
+    // todo: or should we throw?
+  }
+  let unencryptedKeysToSave = await filterKeysToSave(unencryptedPrvs, existingKeys);
+  let encryptedKeys: Key[] = [];
+  if (unencryptedKeysToSave.length) {
+    if (passphrase === undefined) {
+      // trying to find a passphrase that unlocks at least one key
+      const passphrases = await PassphraseStore.getMany(acctEmail, existingKeys);
+      passphrase = passphrases.find(pp => pp !== undefined);
+    }
+    if (passphrase !== undefined) {
+      const pp = passphrase;
```

Comment:
I suppose this line shouldn't be needed.

Reply:
It's a trick needed to make Typescript compiler know this value doesn't change during map (is not `undefined`).
Otherwise, there is an error :(